File: 2D Dynamic Programming/0834. Sum of Distances in Tree/0834.cpp
#pragma GCC optimize('0fast','unroll-loops')

class Solution {
public:
    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);


        unordered_map<int,vector<int>>adjList;
        for(auto& edge : edges){
            adjList[edge[0]].push_back(edge[1]);
            adjList[edge[1]].push_back(edge[0]);
        }

        vector<int>res(n,0);
        vector<int>count(n,1);

        dfs1(0,-1,adjList,res,count);
        dfs2(0,-1,adjList,n,res,count);
        return res;
    }
private:


    void dfs1(int node, int parent, auto& adjList, auto& res, auto& count) {
        for(auto& child : adjList[node]){
            if(child!=parent){
                dfs1(child,node,adjList,res,count);
                count[node]+=count[child];
                res[node]+=res[child]+count[child];
            }
        }
    }

    void dfs2(int node, int parent, auto& adjList, int& n, auto& res, auto& count) {
        for(auto& child : adjList[node]){
            if(child!=parent){
                res[child]=res[node]-count[child]+(n-count[child]);
                dfs2(child,node,adjList,n,res,count);
            }
        }
    }
};

File: 2D Dynamic Programming/0834. Sum of Distances in Tree/0834.py
class Solution:
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        adjList=defaultdict(list)
        for u,v in edges:
            adjList[u].append(v)
            adjList[v].append(u)

        res=[0]*n
        count=[1]*n

        def dfs1(node, parent):
            for child in adjList[node]:
                if child!=parent:
                    dfs1(child,node)
                    count[node]+=count[child]
                    res[node]+=res[child]+count[child]
        
        def dfs2(node,parent):
            for child in adjList[node]:
                if child!=parent:
                    res[child]=res[node]-count[child]+(n-count[child])
                    dfs2(child,node)
        
        return res